rfcs/0007-handling-corruption.md

```diff
@@ -217,108 +209,184 @@ durability and availability systems on top of SlateDB.
+- If the user-defined `DBOptions.on_corruption` callback does not return an
+  error, compaction will ignore the corruption to the best of its ability,
+  which may result in data loss.
+
+Compaction will continue to attempt future runs, aborting each time corruption
+is encountered until the corruption is repaired or compaction is halted. If
+compaction were to never run again, it is possible that the operator might
+remain unaware that compaction has stopped and fail to restart it after
+repairing the corruption.
+
+#### Handling SST Corruption During Get
+If SlateDB detects corruption while searching for a key during a `get` call,
+the `DBOptions.on_corruption` callback is invoked.
+
+- If the callback returns an error, the `get` caller will receive an error that
+  includes the error returned by the callback.
+- If the callback does not return an error, no error will be returned to the
+  `get` caller.
+
+The error returned should be of a distinct type, allowing it to be easily
+identified as either non-transient corruption or a transient error (such as
+network connectivity issues).
+
+Since `get` calls may be ongoing during corruption, it is possible that the
+callback will be called continuously and by multiple threads. Therefore, any
+user-defined callback must be thread-safe. Additionally, the callback should
+avoid slow or blocking operations to ensure the database remains available if
+corruption is encountered.
+
+To support operators who value durability, the developer making the `get` call
+can choose to abort the application if a corruption error is returned. The
+application can then provide operators with detailed information included in
+the error returned by SlateDB, including the nature of the corruption and
+suggestions on how to fix it.
+
+##### Remaining Available
+In the case where corruption is encountered, and `DBOptions.on_corruption` allows
+operation to continue, SlateDB should endeavor to return the most recent 
+valid data it can.
+
+- If the SST Bloom Filter indicates a value exists in the SST, yet the key was
+  not found, and corruption was detected during iteration, the `get` caller
+  should receive a Corruption error.
+- If the most recent value is in a non-corrupt block, the `get` call should
+  return successfully
+- If the SST index is corrupt and `get` has not yet found the KV in an earlier
+  SST, the call should receive a Corruption error. (This effectively prevents
+  reads in SSTs older than the first corrupt SST)
+- MORE?
+
+#### Corruption Inspection and Repair
+To support the repair of SSTs, I propose introducing two new functions:
+`sst_inspect()` and `sst_repair()`. These functions will provide developers with
+tools to inspect the current contents of a specific SST, and repair a specific
+SST that has been identified as corrupt.
+
+To support the repair of the WAL, I propose introducing two additional functions:
+`wal_inspect()` and `wal_repair()`. These functions will allow developers to
```

Comment:
In the Rust impl, we might be able to get away with just `inspect()` and `repair()` since we have:

```rust
pub(crate) enum SsTableId {
    Wal(u64),
    Compacted(Ulid),
}
```

So we could have `fn inspect(id: SsTableId)`. Not sure how y'all have done it on the Go side.

Or is this suggesting you would take no params, and inspect _all_ SSTs? I think we should ask users to provide a specific ID for inspection/repair.

Reply:
Ah, nevermind. I see your notes below about WAL/SST variants accepting raw files and SST IDs. This is exactly what I was after.